[c10d] Integrate NCCL new `ncclAllToAll` into PT

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #164265

NCCL 2.28.3 now supports `ncclAlltoAll`, this PR aims at integrating it. (release note: https://docs.nvidia.com/deeplearning/nccl/release-notes/rel_2-28-3.html#rel_2-28-3)

fduwjj committed:
[c10d] Integrate ncclAllToAll into PT

[ghstack-poisoned]

Skylion007 commented on torch/csrc/cuda/nccl.cpp:
```diff
@@ -850,6 +850,8 @@ void all2all_single_equal_split(
   // operations issued as a part of the collective (e.g. alltoall) vs those
   // inside traditional p2p operations.
   NCCL_CHECK(ncclAllToAll(sendbuff, recvbuff, count, type, comm, stream));
+#elif ((NCCL_MAJOR > 2) || ((NCCL_MAJOR == 2) && (NCCL_MINOR >= 28)))
```
Wait, yeah the API is the same so we can just define it ourselves in OSS and reuse the logic...

fduwjj replied:
the API is different unfortunately...

Skylion007 replied:
Can we at least add a comment to explain it's from FBCODE lol?

Skylion007 replied:
@fduwjj Ah, one is lower case...
   
```NCCL_CHECK(ncclAllToAll(sendbuff, recvbuff, count, type, comm, stream));```
  vs 
```NCCL_CHECK(ncclAlltoAll(sendbuff, recvbuff, count, type, comm, stream));```

fduwjj committed:
Update on "[c10d] Integrate ncclAllToAll into PT"


NCCL 2.28.3 now supports `ncclAlltoAll`, this PR aims at integrating it. (release note: https://docs.nvidia.com/deeplearning/nccl/release-notes/rel_2-28-3.html#rel_2-28-3)

[ghstack-poisoned]

fduwjj commented on torch/csrc/cuda/nccl.cpp:
```diff
@@ -850,6 +850,8 @@ void all2all_single_equal_split(
   // operations issued as a part of the collective (e.g. alltoall) vs those
   // inside traditional p2p operations.
   NCCL_CHECK(ncclAllToAll(sendbuff, recvbuff, count, type, comm, stream));
+#elif ((NCCL_MAJOR > 2) || ((NCCL_MAJOR == 2) && (NCCL_MINOR >= 28)))
```
no this macro is not available in this file..

fduwjj committed:
Update on "[c10d] Integrate ncclAllToAll into PT"


NCCL 2.28.3 now supports `ncclAlltoAll`, this PR aims at integrating it. (release note: https://docs.nvidia.com/deeplearning/nccl/release-notes/rel_2-28-3.html#rel_2-28-3)

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Integrate ncclAllToAll into PT"


NCCL 2.28.3 now supports `ncclAlltoAll`, this PR aims at integrating it. (release note: https://docs.nvidia.com/deeplearning/nccl/release-notes/rel_2-28-3.html#rel_2-28-3)

[ghstack-poisoned]

fduwjj commented:
@pytorchbot -i merge
